### PR TITLE
Additional redirect to handle multi-dns configuration

### PR DESCRIPTION
**Pull request by mrajashree**

### CODE REVIEW

File: service/route_handlers.go [hunk 1]
```diff
@@ -473,9 +482,24 @@ func getPossibleRequestIDs(r *http.Request, s *model.RancherSamlServiceProvider)
...
+		if err != nil {
+			log.Errorf("Error in parsing query params")
+			return []string{}, "", "", fmt.Errorf("Error in parsing query params: %v", err)
+		}
+		redirectBackBaseValue = parsedQuery.Get(redirectBackBase)
+		redirectBackPathValue = parsedQuery.Get(redirectBackPath)
+		log.Debugf("redirectBackBaseValue, redirectBackPathValue from claims: %v, %v", redirectBackBaseValue, redirectBackPathValue)
 	}
-	return rv
+	return reqIDs, redirectBackBaseValue, redirectBackPathValue, nil
```
Comment (prachidamle) [inline service/route_handlers.go:520]: Can you describe this change? Why are we reading the redirectBackBaseValue/redirectBackPathValue from SamlAssertion? Where were we reading it earlier and why this change is needed?

Comment (mrajashree) [reply service/route_handlers.go:520]: Prior to refactor, saml library's Authorize method would redirect to request URI picked from relayState. And then HandleSamlPost method would get redirectBackBase & Path from the request's query parameters, and do a redirect to v1-auth/saml/authtoken.

So now, this code will do what Authorize method did, and get the relayState and redirectBackBase/Path from that and send it. I've added a comment above in that function

File: service/route_handlers.go [hunk 2]
```diff
@@ -546,6 +574,59 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
...
+
+	/* token is created at this point. If the redirectBackBase is not same as the current URL's domain, then we cannot set the token as cookie
+	yet. So we will first send a 200 POST to /v1-auth/saml/tokenhtml, which will contain token cookie and redirectBackBase in HTML body,
+	and JS which will submit form on load to redirect URL. The handler for v1-auth/saml/tokenhtml will then set the token as cookie
+	*/
+
+	if redirectBackBaseValue != r.URL.Host {
+		newRedirectURL := redirectBackBaseValue + postSamlTokenHTML
+		w.Header().Add("Content-type", "text/html")
+		tpl := `
```
Comment (cjellick) [inline service/route_handlers.go]: you might as well make this a constant

File: service/route_handlers.go [hunk 3]
```diff
@@ -448,15 +457,15 @@ func ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		HandleSamlAssertion(w, r, assertion, server.SamlServiceProvider, redirectBackBaseValue, redirectBackPathValue)
 		return
 	}
 
 	http.NotFoundHandler().ServeHTTP(w, r)
 }
 
-func getPossibleRequestIDs(r *http.Request, s *model.RancherSamlServiceProvider) []string {
-	rv := []string{}
+func getPossibleRequestIDs(r *http.Request, s *model.RancherSamlServiceProvider) (reqIDs []string, redirectBackBaseValue string,
```
Comment (cjellick) [inline service/route_handlers.go]: getPossibleRequestIDs does not accurately reflect what this function does because it also returns the redirect pieces

Comment (cjellick) [reply service/route_handlers.go]: why have you named the return parameters? My opinion, for clarity you shouldnt do that unless you really need to

File: service/route_handlers.go [hunk 4]
```diff
@@ -473,9 +482,29 @@ func getPossibleRequestIDs(r *http.Request, s *model.RancherSamlServiceProvider)
...
+		if err != nil {
+			log.Errorf("Error in getting query params")
+			return []string{}, "", "", fmt.Errorf("Error in getting query params: %v", err)
+		}
+		parsedQuery, err := url.ParseQuery(query.RawQuery)
+		if err != nil {
+			log.Errorf("Error in parsing query params")
+			return []string{}, "", "", fmt.Errorf("Error in parsing query params: %v", err)
+		}
+		redirectBackBaseValue = parsedQuery.Get(redirectBackBase)
```
Comment (cjellick) [inline service/route_handlers.go:516]: Do you have fallback logic anywhere that will allow things to work if the redirect parameters are not on the URL? If not, I think you should add that

File: service/route_handlers.go [hunk 5]
```diff
@@ -546,6 +574,59 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
 	if XForwardedProtoValue == "https" {
 		secure = true
 	}
+
+	/* token is created at this point. If the redirectBackBase is not same as the current URL's domain, then we cannot set the token as cookie
+	yet. So we will first send a 200 POST to /v1-auth/saml/tokenhtml, which will contain token cookie and redirectBackBase in HTML body,
+	and JS which will submit form on load to redirect URL. The handler for v1-auth/saml/tokenhtml will then set the token as cookie
+	*/
+
+	if redirectBackBaseValue != r.URL.Host {
```
Comment (cjellick) [inline service/route_handlers.go]: Does redirectBackBaseValue have the http:// portion in it? If so, will it always be different from r.URL.Host?

File: service/route_handlers.go [hunk 6]
```diff
@@ -554,25 +635,83 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
...
 	http.Redirect(w, r, redirectURL, http.StatusFound)
 	return
 }
 
+func PostSamlTokenHTML(w http.ResponseWriter, r *http.Request) {
+	var token, finalRedirectURL string
+	var secure bool
+	var err error
+
+	r.ParseForm()
```
Comment (cjellick) [inline service/route_handlers.go]: i think youre swallowing an error here

File: service/route_handlers.go [hunk 7]
```diff
@@ -554,25 +635,83 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
...
+
+	if len(r.Form["token"]) > 0 { //array of string
```
Comment (cjellick) [inline service/route_handlers.go]: best to add panic protection to all these key based lookups. something like
```
if t, ok := r.Form["token"]; ok && len(t) {
```

File: service/route_handlers.go [hunk 8]
```diff
@@ -554,25 +635,83 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
...
+
+	if len(r.Form["finalRedirectURL"]) > 0 {
+		finalRedirectURL = r.Form["finalRedirectURL"][0]
+	}
+
+	if len(r.Form["secure"]) > 0 {
+		secure, err = strconv.ParseBool(r.Form["secure"][0])
+		if err != nil {
+			log.Errorf("PostSamlTokenHTML: Error in parsing secure flag for token: %v", err)
+			return
```
Comment (cjellick) [inline service/route_handlers.go]: what would happen to a request if you hit this code block?
Seems like you would just return a blank 200 response? if so, that doesnt seem right. seems like it should 500

File: service/route_handlers.go (hunk 3 above)
Comment (mrajashree) [reply service/route_handlers.go]: I thought I'll just name them since they contain an array of strings and 2 strings.

File: service/route_handlers.go (hunk 1 above)
Comment (prachidamle) [reply service/route_handlers.go:520]: I think we need to remove the code that was setting redirectBackBase/Path to the global serviceProvider struct earlier

File: service/route_handlers.go (hunk 3 above)
Comment (prachidamle) [reply service/route_handlers.go]: @cjellick any reason from 'go' perspective we should not be naming the return fields? Atleast these made it easy for me to read the code.

File: service/route_handlers.go [hunk 9]
```diff
@@ -554,25 +635,83 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
...
 			}
+
+			if strings.Contains(w, ":") {
+				d := strings.SplitN(w, ":", 2)
+				log.Debugf("Domain: %v", d[0])
+				if d[0] == redirectBackBaseDomain || d[0] == redirectBackBaseDomainNonPort {
+					log.Debugf("Whitelisted domain matched: %v", d[0])
+					return true
+				}
+			}
```
Comment (prachidamle) [inline service/route_handlers.go]: Since we are expecting that Admin can enter the whitelisted string as just a domain x.com OR complete url "x.com:port" - we should check if the whitelisted value has port and make sure it matches the port presented by 'redirectBackBaseDomain'.

If whitelisted value is just a domain, we can check only the domain matches irrespective of the port presented in 'redirectBackBaseDomain'

Comment (prachidamle) [reply service/route_handlers.go]: Basically we need to ensure the whitelisted value matches as-is.

File: service/route_handlers.go (hunk 4 above)
Comment (mrajashree) [reply service/route_handlers.go:516]: Added the logic in HandleSamlAssertion

**Comment by cjellick** [comment]
LGTM
